File: java/src/main/java/io/trino/velox/NativeConfigs.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.trino.velox;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.inject.Inject;
import io.trino.execution.TaskManagerConfig;
import io.trino.operator.DirectExchangeClientConfig;

public class NativeConfigs
{
    private final long maxOutputPageBytes;
    private final int maxWorkerThreads;
    private final int maxDriversPerTask;
    private final int taskConcurrency;
    private final int exchangeClientThreads;
    private final long queryMaxMemoryPerNode;
    private final String logVerboseModules;
    private final long maxNodeMemory;
    private final boolean useMmapAllocator;
    private final boolean useMmapArena;
    private final int mmapArenaCapacityRatio;
    private final boolean asyncDataCacheEnabled;
    private final long asyncCacheSsdSize;
    private final long asyncCacheSsdCheckpointSize;
    private final boolean asyncCacheSsdDisableFileCow;
    private final String asyncCacheSsdPath;
    private final boolean enableMemoryLeakCheck;
    private final boolean enableMemoryArbitration;
    private final int reservedMemoryPoolCapacityPercentage;
    private final long initMemoryPoolCapacity;
    private final long minMemoryPoolTransferCapacity;
    private final int maxHttpSessionReadBufferSize;
    private final boolean joinSpillEnabled;
    private final boolean aggSpillEnabled;
    private final boolean orderBySpillEnabled;
    private final boolean spillEnabled;
    private final String spillDir;
    private final long aggregationSpillMemoryThreshold;
    private final long joinSpillMemoryThreshold;
    private final long orderBySpillMemoryThreshold;

    @Inject
    public NativeConfigs(
            TaskManagerConfig taskManagerConfig,
            DirectExchangeClientConfig directExchangeClientConfig,
            NativeTaskConfig nativeTaskConfig)
    {
        this(directExchangeClientConfig.getMaxResponseSize().toBytes(),
                taskManagerConfig.getMaxWorkerThreads(),
                taskManagerConfig.getMaxDriversPerTask(),
                taskManagerConfig.getTaskConcurrency(),
                directExchangeClientConfig.getClientThreads(),
                nativeTaskConfig.getMaxQueryMemoryPerNode().toBytes(),
                nativeTaskConfig.getLogVerboseModules(),
                nativeTaskConfig.getMaxNodeMemory().toBytes(),
                nativeTaskConfig.isUseMmapAllocator(),
                nativeTaskConfig.isUseMmapArena(),
                nativeTaskConfig.getMmapArenaCapacityRatio(),
                nativeTaskConfig.isAsyncDataCacheEnabled(),
                nativeTaskConfig.getAsyncCacheSsdSize().toBytes(),
                nativeTaskConfig.getAsyncCacheSsdCheckpointSize().toBytes(),
                nativeTaskConfig.isAsyncCacheSsdDisableFileCow(),
                nativeTaskConfig.getAsyncCacheSsdPath(),
                nativeTaskConfig.isEnableMemoryLeakCheck(),
                nativeTaskConfig.isEnableMemoryArbitration(),
                nativeTaskConfig.getReservedMemoryPoolCapacityPercentage(),
                nativeTaskConfig.getInitMemoryPoolCapacity(),
                nativeTaskConfig.getMinMemoryPoolTransferCapacity(),
                nativeTaskConfig.getMaxHttpSessionReadBufferSize(),
                nativeTaskConfig.isSpillEnabled(),
                nativeTaskConfig.isJoinSpillEnabled(),
                nativeTaskConfig.isAggSpillEnabled(),
                nativeTaskConfig.isOrderBySpillEnabled(),
                nativeTaskConfig.getSpillDir(),
                nativeTaskConfig.getJoinSpillMemoryThreshold(),
                nativeTaskConfig.getAggregationSpillMemoryThreshold(),
                nativeTaskConfig.getOrderBySpillMemoryThreshold());
    }

    @JsonCreator
    public NativeConfigs(
            @JsonProperty long maxOutputPageBytes,
            @JsonProperty int maxWorkerThreads,
            @JsonProperty int maxDriversPerTask,
            @JsonProperty int taskConcurrency,
            @JsonProperty int exchangeClientThreads,
            @JsonProperty long queryMaxMemoryPerNode,
            @JsonProperty String logVerboseModules,
            @JsonProperty long maxNodeMemory,
            @JsonProperty boolean useMmapAllocator,
            @JsonProperty boolean useMmapArena,
            @JsonProperty int mmapArenaCapacityRatio,
            @JsonProperty boolean asyncDataCacheEnabled,
            @JsonProperty long asyncCacheSsdSize,
            @JsonProperty long asyncCacheSsdCheckpointSize,
            @JsonProperty boolean asyncCacheSsdDisableFileCow,
            @JsonProperty String asyncCacheSsdPath,
            @JsonProperty boolean enableMemoryLeakCheck,
            @JsonProperty boolean enableMemoryArbitration,
            @JsonProperty int reservedMemoryPoolCapacityPercentage,
            @JsonProperty long initMemoryPoolCapacity,
            @JsonProperty long minMemoryPoolTransferCapacity,
            @JsonProperty int maxHttpSessionReadBufferSize,
            @JsonProperty boolean spillEnabled,
            @JsonProperty boolean joinSpillEnabled,
            @JsonProperty boolean aggSpillEnabled,
            @JsonProperty boolean orderBySpillEnabled,
            @JsonProperty String spillDir,
            @JsonProperty long joinSpillMemoryThreshold,
            @JsonProperty long aggregationSpillMemoryThreshold,
            @JsonProperty long orderBySpillMemoryThreshold)
    {
        this.maxOutputPageBytes = maxOutputPageBytes;
        this.maxWorkerThreads = maxWorkerThreads;
        this.maxDriversPerTask = maxDriversPerTask;
        this.taskConcurrency = taskConcurrency;
        this.exchangeClientThreads = exchangeClientThreads;
        this.queryMaxMemoryPerNode = queryMaxMemoryPerNode;
        this.logVerboseModules = logVerboseModules;
        this.maxNodeMemory = maxNodeMemory;
        this.useMmapAllocator = useMmapAllocator;
        this.useMmapArena = useMmapArena;
        this.mmapArenaCapacityRatio = mmapArenaCapacityRatio;
        this.asyncDataCacheEnabled = asyncDataCacheEnabled;
        this.asyncCacheSsdSize = asyncCacheSsdSize;
        this.asyncCacheSsdCheckpointSize = asyncCacheSsdCheckpointSize;
        this.asyncCacheSsdDisableFileCow = asyncCacheSsdDisableFileCow;
        this.asyncCacheSsdPath = asyncCacheSsdPath;
        this.enableMemoryLeakCheck = enableMemoryLeakCheck;
        this.enableMemoryArbitration = enableMemoryArbitration;
        this.reservedMemoryPoolCapacityPercentage = reservedMemoryPoolCapacityPercentage;
        this.initMemoryPoolCapacity = initMemoryPoolCapacity;
        this.minMemoryPoolTransferCapacity = minMemoryPoolTransferCapacity;
        this.maxHttpSessionReadBufferSize = maxHttpSessionReadBufferSize;
        this.spillEnabled = spillEnabled;
        this.joinSpillEnabled = joinSpillEnabled;
        this.aggSpillEnabled = aggSpillEnabled;
        this.orderBySpillEnabled = orderBySpillEnabled;
        this.spillDir = spillDir;
        this.joinSpillMemoryThreshold = joinSpillMemoryThreshold;
        this.aggregationSpillMemoryThreshold = aggregationSpillMemoryThreshold;
        this.orderBySpillMemoryThreshold = orderBySpillMemoryThreshold;
    }

    @JsonProperty
    public long getMaxOutputPageBytes()
    {
        return maxOutputPageBytes;
    }

    @JsonProperty
    public int getMaxWorkerThreads()
    {
        return maxWorkerThreads;
    }

    @JsonProperty
    public int getMaxDriversPerTask()
    {
        return maxDriversPerTask;
    }

    @JsonProperty
    public int getTaskConcurrency()
    {
        return taskConcurrency;
    }

    @JsonProperty
    public int getExchangeClientThreads()
    {
        return exchangeClientThreads;
    }

    @JsonProperty
    public long getQueryMaxMemoryPerNode()
    {
        return queryMaxMemoryPerNode;
    }

    @JsonProperty
    public String getLogVerboseModules()
    {
        return logVerboseModules;
    }

    @JsonProperty
    public long getMaxNodeMemory()
    {
        return maxNodeMemory;
    }

    @JsonProperty
    public boolean isUseMmapAllocator()
    {
        return useMmapAllocator;
    }

    @JsonProperty
    public boolean isUseMmapArena()
    {
        return useMmapArena;
    }

    @JsonProperty
    public int getMmapArenaCapacityRatio()
    {
        return mmapArenaCapacityRatio;
    }

    @JsonProperty
    public boolean isAsyncDataCacheEnabled()
    {
        return asyncDataCacheEnabled;
    }

    @JsonProperty
    public long getAsyncCacheSsdSize()
    {
        return asyncCacheSsdSize;
    }

    @JsonProperty
    public long getAsyncCacheSsdCheckpointSize()
    {
        return asyncCacheSsdCheckpointSize;
    }

    @JsonProperty
    public boolean isAsyncCacheSsdDisableFileCow()
    {
        return asyncCacheSsdDisableFileCow;
    }

    @JsonProperty
    public String getAsyncCacheSsdPath()
    {
        return asyncCacheSsdPath;
    }

    @JsonProperty
    public boolean isEnableMemoryLeakCheck()
    {
        return enableMemoryLeakCheck;
    }

    @JsonProperty
    public boolean isEnableMemoryArbitration()
    {
        return enableMemoryArbitration;
    }

    @JsonProperty
    public int getReservedMemoryPoolCapacityPercentage()
    {
        return reservedMemoryPoolCapacityPercentage;
    }

    @JsonProperty
    public long getInitMemoryPoolCapacity()
    {
        return initMemoryPoolCapacity;
    }

    @JsonProperty
    public long getMinMemoryPoolTransferCapacity()
    {
        return minMemoryPoolTransferCapacity;
    }

    @JsonProperty
    public int getMaxHttpSessionReadBufferSize()
    {
        return maxHttpSessionReadBufferSize;
    }

    @JsonProperty
    public boolean getSpillEnabled()
    {
        return spillEnabled;
    }

    @JsonProperty
    public boolean getJoinSpillEnabled()
    {
        return joinSpillEnabled;
    }

    @JsonProperty
    public boolean getAggSpillEnabled()
    {
        return aggSpillEnabled;
    }

    @JsonProperty
    public boolean getOrderBySpillEnabled()
    {
        return orderBySpillEnabled;
    }

    @JsonProperty
    public String getSpillDir()
    {
        return spillDir;
    }

    @JsonProperty
    public long getAggregationSpillMemoryThreshold()
    {
        return aggregationSpillMemoryThreshold;
    }

    @JsonProperty
    public long getJoinSpillMemoryThreshold()
    {
        return joinSpillMemoryThreshold;
    }

    @JsonProperty
    public long getOrderBySpillMemoryThreshold()
    {
        return orderBySpillMemoryThreshold;
    }
}

File: cpp/src/NativeConfigs.cpp
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "src/NativeConfigs.h"

#include "src/protocol/external/json/json.hpp"
#include "velox/core/Config.h"
#include "velox/core/QueryConfig.h"

using namespace facebook::velox;

namespace io::trino::bridge {

#define GET_KEY_FROM_JSON(KEY, JSON) \
  if (JSON.contains(#KEY)) {         \
    JSON.at(#KEY).get_to(KEY);       \
  }

NativeConfigs& NativeConfigs::instance() {
  static NativeConfigs configs;
  return configs;
}

void NativeConfigs::initialize(const std::string& configJsonString) {
  if (initialized) {
    return;
  }
  VLOG(2) << "Initializing Native Configs with config json: " << configJsonString;

  nlohmann::json configJson = nlohmann::json::parse(configJsonString);
  GET_KEY_FROM_JSON(maxOutputPageBytes, configJson);
  GET_KEY_FROM_JSON(maxWorkerThreads, configJson);
  GET_KEY_FROM_JSON(maxDriversPerTask, configJson);
  GET_KEY_FROM_JSON(taskConcurrency, configJson);
  GET_KEY_FROM_JSON(exchangeClientThreads, configJson);
  GET_KEY_FROM_JSON(queryMaxMemoryPerNode, configJson);
  GET_KEY_FROM_JSON(maxNodeMemory, configJson);
  GET_KEY_FROM_JSON(useMmapAllocator, configJson);
  GET_KEY_FROM_JSON(useMmapArena, configJson);
  GET_KEY_FROM_JSON(mmapArenaCapacityRatio, configJson);
  GET_KEY_FROM_JSON(asyncDataCacheEnabled, configJson);
  GET_KEY_FROM_JSON(asyncCacheSsdSize, configJson);
  GET_KEY_FROM_JSON(asyncCacheSsdCheckpointSize, configJson);
  GET_KEY_FROM_JSON(asyncCacheSsdDisableFileCow, configJson);
  GET_KEY_FROM_JSON(asyncCacheSsdPath, configJson);
  GET_KEY_FROM_JSON(enableMemoryLeakCheck, configJson);
  GET_KEY_FROM_JSON(enableMemoryArbitration, configJson);
  GET_KEY_FROM_JSON(memoryArbitratorKind, configJson);
  GET_KEY_FROM_JSON(reservedMemoryPoolCapacityPercentage, configJson);
  GET_KEY_FROM_JSON(initMemoryPoolCapacity, configJson);
  GET_KEY_FROM_JSON(minMemoryPoolTransferCapacity, configJson);
  GET_KEY_FROM_JSON(maxHttpSessionReadBufferSize, configJson);
  GET_KEY_FROM_JSON(spillEnabled, configJson);
  GET_KEY_FROM_JSON(joinSpillEnabled, configJson);
  GET_KEY_FROM_JSON(aggSpillEnabled, configJson);
  GET_KEY_FROM_JSON(orderBySpillEnabled, configJson);
  GET_KEY_FROM_JSON(spillDir, configJson);
  GET_KEY_FROM_JSON(joinSpillMemoryThreshold, configJson);
  GET_KEY_FROM_JSON(aggregationSpillMemoryThreshold, configJson);
  GET_KEY_FROM_JSON(orderBySpillMemoryThreshold, configJson);
  GET_KEY_FROM_JSON(concurrentLifespans, configJson);
  GET_KEY_FROM_JSON(baseUrl, configJson);
  GET_KEY_FROM_JSON(instanceId, configJson);
  GET_KEY_FROM_JSON(httpMaxAllocateBytes, configJson);
  GET_KEY_FROM_JSON(httpsClientCertAndKeyPath, configJson);
  GET_KEY_FROM_JSON(httpsSupportedCiphers, configJson);

  if (configJson.contains("logVerboseModules")) {
    std::string _logVerboseModules;
    configJson.at("logVerboseModules").get_to(_logVerboseModules);

    if (_logVerboseModules.back() != ',') {
      _logVerboseModules.push_back(',');
    }

    size_t start = 0;
    size_t index = 0;
    while (index < _logVerboseModules.size()) {
      while (index < _logVerboseModules.size() && _logVerboseModules[index] != ',')
        ++index;

      if (index >= _logVerboseModules.size()) break;

      if (index - start < 3) {
        // There are at least "m=l" for 3 characters.
        start = ++index;
        continue;
      }

      const std::string& logVerboseModule =
          _logVerboseModules.substr(start, index - start);
      size_t pos = logVerboseModule.find('=');
      if (pos == start || pos + 1 == index) {
        start = ++index;
        continue;
      }

      const std::string& moduleName = logVerboseModule.substr(0, pos);
      int level =
          std::stoi(logVerboseModule.substr(pos + 1, logVerboseModule.size() - pos - 1));

      logVerboseModules[moduleName] = level;

      start = ++index;
    }
  }

  initialized = true;
}

std::unordered_map<std::string, std::string> NativeConfigs::getQueryConfigs() const {
  return {
      {core::QueryConfig::kMaxPartitionedOutputBufferSize,
       std::to_string(getMaxOutputPageBytes())},
      {core::QueryConfig::kSpillEnabled, std::to_string(getSpillEnabled())},
      {core::QueryConfig::kJoinSpillEnabled, std::to_string(getJoinSpillEnabled())},
      {core::QueryConfig::kAggregationSpillEnabled, std::to_string(getAggSpillEnabled())},
      {core::QueryConfig::kOrderBySpillEnabled, std::to_string(getOrderBySpillEnabled())},
      {core::QueryConfig::kJoinSpillMemoryThreshold,
       std::to_string(getJoinSpillMemoryThreshold())},
      {core::QueryConfig::kAggregationSpillMemoryThreshold,
       std::to_string(getAggregationSpillMemoryThreshold())},
      {core::QueryConfig::kOrderBySpillMemoryThreshold,
       std::to_string(getJoinSpillMemoryThreshold())}};
}

std::unordered_map<std::string, std::shared_ptr<Config>>
NativeConfigs::getConnectorConfigs() {
  return {};
}

}  // namespace io::trino::bridge
